modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -4629,3 +4636,179 @@ def from_dataframe(cls, df: "ProtocolDataframe") -> "PandasDataframe":
+        is_trivial_idx = None
+
+        def copartition(df, fill_value):
+            nonlocal is_trivial_idx
+            copartition = False
+            if is_trivial_idx is None:
+                is_trivial_idx = is_trivial_index(self.index)
+            if (
+                is_trivial_idx != is_trivial_index(df.index)
+                or (not is_trivial_idx and any(self.index != df.index))
+                or (parts_len < len(df._partitions))
+            ):
+                copartition = True
+            else:
+                nonlocal parts, lengths
+                if parts is None:
+                    parts = [p[0] for p in self._partitions]
+                    lengths = self._get_lengths(parts, Axis.ROW_WISE)
+
+                df_parts = [p[0] for p in df._partitions]
+                df_part_lengths = df._get_lengths(df_parts, Axis.ROW_WISE)
+                if any(lengths[i] != df_part_lengths[i] for i in range(len(df_parts))):
+                    copartition = True
+            if copartition:
+                _, list_of_right_parts, joined_index, row_lengths = self._copartition(
+                    Axis.ROW_WISE.value,
+                    df,
+                    how="left",
+                    sort=False,
+                    fill_value=fill_value,
+                )
+                df = self.__constructor__(
+                    list_of_right_parts[0],
+                    joined_index,
+                    df.columns,
+                    row_lengths,
+                    df.column_widths,
+                )
+            return df
+
+        use_map = single_wrap
+        new_caselist = []
+        for condition, replacement in caselist:
+            if callable(condition):
+                if single_wrap:
+                    condition = wrapper_put(condition)
+            else:
+                use_map = False
+                if isinstance(condition, cls):
+                    condition = copartition(condition, True)
+            if callable(replacement):
+                if single_wrap:
+                    replacement = wrapper_put(replacement)
+            elif use_map and is_list_like(replacement):
+                use_map = False
+                if isinstance(replacement, cls):
+                    replacement = copartition(replacement, None)
+            new_caselist.append((condition, replacement))
+
+        # If all the conditions are callable and the replacements are either
+        # callable or scalar, use map().
+        if use_map:
+            return self.map(func=remote_fn, func_args=[new_caselist], lazy=True)
+
+        def map_data(
+            part_offset,
+            part_len,
+            data,
+            data_offset,
+            fill_value,
+        ):
+            if isinstance(data, cls):
+                if part_offset < len(data._partitions):
+                    return data._partitions[part_offset][0]._data
+                else:
+                    return [fill_value] * part_len
```

Comment:
Does this ever happen? I though `._copartition()` should have aligned partitioning

Reply:
ping